p2p/ipld/read.go

```diff
@@ -30,23 +32,32 @@ func RetrieveBlockData(
 ) (types.Data, error) {
 	edsWidth := len(dah.RowsRoots)
 	sc := newshareCounter(ctx, uint32(edsWidth))
+	sem := make(chan interface{}, maxGoRoutines)
 
 	// convert the row and col roots into Cids
 	rowRoots := dah.RowsRoots.Bytes()
 	colRoots := dah.ColumnRoots.Bytes()
 
 	// sample 1/4 of the total extended square by sampling half of the leaves in
 	// half of the rows
-	for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-		for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
-			rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
-			if err != nil {
-				return types.Data{}, err
+	go func() {
+		for _, row := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+			for _, col := range uniqueRandNumbers(edsWidth/2, edsWidth) {
+				rootCid, err := plugin.CidFromNamespacedSha256(rowRoots[row])
+				if err != nil {
+					select {
+					case <-sc.ctx.Done():
+					case sc.errc <- err:
+					}
+				}
+				sem <- struct{}{}
+				go func(rootCid cid.Cid, row uint32, col uint32) {
+					sc.retrieveShare(rootCid, true, row, col, dag)
+					<-sem
+				}(rootCid, row, col)
 			}
-
-			go sc.retrieveShare(rootCid, true, row, col, dag)
 		}
-	}
+	}()
```

Comment:
I think we will want to `defer close(sem)` in the beginning of this goroutine, and then return if the context is finished on line 49, just to wrap up loose ends in the case of a prematurely canceled context